controller/orphan_controller.go

```diff
@@ -371,6 +572,55 @@ func (oc *OrphanController) updateConditions(orphan *longhorn.Orphan) error {
 	return nil
 }
 
+// If the InstanceState condition status is true, the corresponding instance is still exist in instance manager
+func (oc *OrphanController) updateInstanceStateCondition(orphan *longhorn.Orphan, instanceType longhorn.InstanceType) (err error) {
+	var instanceState longhorn.InstanceState
+	defer func() {
+		if err != nil {
+			return
+		}
+
+		status := longhorn.ConditionStatusTrue
+		if instanceState == longhorn.InstanceStateTerminated {
+			status = longhorn.ConditionStatusFalse
+		}
+
+		orphan.Status.Conditions = types.SetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeInstanceState, status, string(instanceState), "")
+	}()
+
+	instanceName, instanceManager, err := oc.extractOrphanedInstanceInfo(orphan)
+	im, err := oc.ds.GetInstanceManager(instanceManager)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			oc.logger.WithError(err).Infof("No instance manager for node %v for update instance state of orphan instance %v", oc.controllerID, orphan.Name)
+			instanceState = longhorn.InstanceStateTerminated
+			return nil
+		}
+		return err
+	}
+	imc, err := engineapi.NewInstanceManagerClient(im, false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to check node down or missing manager for node %v", orphan.Spec.NodeID)
+	}
+	defer func() {
+		if closeErr := imc.Close(); closeErr != nil {
+			oc.logger.WithError(closeErr).Error("failed to close instance manager client")
+		}
+	}()
+
+	instanceProc, err := imc.InstanceGet(orphan.Spec.DataEngine, instanceName, string(instanceType))
+	switch {
+	case err != nil:
+		return errors.Wrapf(err, "failed to get instance %v", instanceName)
+	case instanceProc == nil:
+		instanceState = longhorn.InstanceStateTerminated
+	default:
+		instanceState = instanceProc.Status.State
+	}
```

Comment:
What's the purpose of reading the instance state?

Reply:
To track the state of corresponding instance, and make trouble shooting easier. User don't need to compare the orphan object with the instance list in im CR.